mPowerSDK/APHScoring.m

```diff
@@ -87,20 +104,287 @@ - (NSDictionary *)summarizeDataset:(NSDictionary *)dataset period:(APHTimelineGr
+
+	for (NSDictionary *dataPoint in [self.dataPoints copy]) {
+        NSArray *rawDataPoints = dataPoint[@"datasetRawDataPoints"];
+        
+        for (NSString *activityTimingChoiceString in self.activityTimingChoicesStrings) {
+            NSMutableDictionary *filteredDataPoint = [dataPoint mutableCopy];
+            
+            NSPredicate *filterPredicate;
+            if ([activityTimingChoiceString isEqualToString:noMedicationTimingKey]) {
+                filterPredicate = [NSPredicate predicateWithFormat:@"(datasetTaskResult == nil) || (datasetTaskResult.MedicationActivityTiming == nil)"];
+            } else {
+                filterPredicate = [NSPredicate predicateWithFormat:@"(datasetTaskResult.MedicationActivityTiming == %@)", activityTimingChoiceString];
+            }
+            
+            NSArray *filteredRawDataPoints = [rawDataPoints filteredArrayUsingPredicate:filterPredicate];
+            filteredDataPoint[@"datasetRawDataPoints"] = filteredRawDataPoints;
+            
+            if (filteredRawDataPoints.count == 0 && ![activityTimingChoiceString isEqualToString:noMedicationTimingKey]) {
+                filteredDataPoint[kDatasetValueKey] = @(NSNotFound);
+            } else if (filteredRawDataPoints.count > 0) {
+                NSNumber *averageRawDataPointValue = [[filteredRawDataPoints valueForKey:kDatasetValueKey] valueForKeyPath:@"@avg.intValue"];
+                filteredDataPoint[kDatasetValueKey] = averageRawDataPointValue;
+            }
+            
+            [self.medTimingDataPoints[activityTimingChoiceString] addObject:filteredDataPoint];
+        }
+	}
+    
+    NSMutableArray *averagedDataPoints = [NSMutableArray new];
+    for (NSDictionary *dataPoint in [self.dataPoints copy]) {
+        NSMutableDictionary *copiedDataPoint = [dataPoint mutableCopy];
+        
+        NSMutableArray *newRawDataPoints = [NSMutableArray new];
+        for (NSString *medTimingChoiceString in self.activityTimingChoicesStrings) {
+            NSArray *points = self.medTimingDataPoints[medTimingChoiceString];
+            NSPredicate *filterPredicate = [NSPredicate predicateWithFormat:@"datasetDateKey == %@", dataPoint[kDatasetDateKey]];
+            NSDictionary *point = [points filteredArrayUsingPredicate:filterPredicate].firstObject;
+            
+            if (!point) {
+                continue;
+            }
+            
+            NSMutableDictionary *pointCopy = [point mutableCopy];
+            [pointCopy removeObjectForKey:kDatasetRawDataPointsKey];
+            pointCopy[kDatasetTaskResultKey] = @{ @"MedicationActivityTiming": medTimingChoiceString };
+            [newRawDataPoints addObject:pointCopy];
+        }
+        
+        copiedDataPoint[kDatasetRawDataPointsKey] = newRawDataPoints;
+        [averagedDataPoints addObject:copiedDataPoint];
+    }
+    
+    self.averagedDataPoints = [averagedDataPoints copy];
+    
+    if (!self.correlatedScoring) { return; }
```

Comment:
Danger, Will Robinson.  An Exit in the middle of a method indicates that the method should be redesigned. Please refactor this to pull out the code before and after into a separate method.